### DISCUSSION TITLE
executionPlan, taskGraph

### DISCUSSION THREAD

**Discussion by aaronms1** [beehive-lab/TornadoVM#266] [Ideas/Proposals]
Hello I'm Aaron a backend-engineer with a strong background in cyberSecurity. I'm interested in using tornadoVM in my encrypted data/transport framework.  Here is my issue/question. Following the KernelApi examples, I found the executionPlan, taskGraphwill not work with my encryption/decryption/transport classes... is there a way i can use the existing executionPlan, taskGraph packages/classes to achieve what I'm attempting to to do, or do I need to write my own 'plan, and graph' packages/classes??
Also I'd love to contribute in any way I can to tornadoVM, and am open to collaborators in my framework.

**Comment by jjfumero** [comment]
Hi @aaronms1 . Thank you for your interest in TornadoVM. 
Can you be more specific about why the current APIs are not enough? Is there anything missing you would like to see (e.g., types, operations, etc). ? We have a set of examples and benchmarks in the TornadoVM suite that can help to get ideas about how to use TornadoVM:
- Examples: https://github.com/beehive-lab/TornadoVM/tree/master/tornado-examples/src/main/java/uk/ac/manchester/tornado/examples 
- Benchmarks: https://github.com/beehive-lab/TornadoVM/tree/master/tornado-benchmarks/src/main/java/uk/ac/manchester/tornado/benchmarks 
- TornadoVM-Examples Repo: https://github.com/jjfumero/tornadovm-examples 

Another great way to learn is through the Unit-tests, in which we test each supported and non-supported feature:
- https://github.com/beehive-lab/TornadoVM/tree/master/tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests 

Besides, we have the documentation webpage in which we explain the Task-Graphs, Execution Plans and how to express parallelism within  TornadoVM: https://tornadovm.readthedocs.io/en/latest/programming.html 

Feel free to file new questions/proposals. 

Feel free to file new questions and proposals.

**Reply by aaronms1** [reply]
Thankyou for the speedy reply. Whats unclear to me is how do i use the api in conjunction with my encryption classes. Do i simply add them as a "T" parameter. exe.. MyCipherClass T. In place of the example Matrix A, Matrix B... parameters?

**Reply by stratika** [reply]
Hi @aaronms1, it is not very clear to me the information about why the API cannot be used. The info about integrating TornadoVM in other projects is available [here](https://github.com/beehive-lab/TornadoVM#5-how-to-use-tornadovm-in-your-projects). Is there any other reason that you believe that the API cannot be used in your encryption classes?

Additionally, there are several examples, such as the [TornadoVM-Ray-Tracer](https://github.com/beehive-lab/TornadoVM-Ray-Tracer) which is an external project that uses the TornadoVM API (TaskGraph, ExecutionPlan) to access GPUs. The way to do so, is by adding the Tornado API as a maven dependency in the pom.xml file. See here (lines 32-35): https://github.com/beehive-lab/TornadoVM-Ray-Tracer/blob/master/pom.xml

**Reply by aaronms1** [reply]
thank you, I will go over the docs again. I must have missed something. Feel free to delete this conversation. I apologize for any inconvenience.

**Reply by stratika** [reply]
No worries, feel free to ask questions! You are more than welcome to give feedback so that we can improve the documentation, etc.

**Comment by aaronms1** [comment]
To be honest, it's probably my fault for not exploring the docs more, but I never found the 'full' page description link that you recently shared with me. I found it now at the bottom of the 'example' snippets on the main README.md.
Once I accomplish my goal with TornadoVm, Ill be sure to share my findings, result, and documentation. I have been working on my encryption framework a little over a year, and am at the point of utilizing Tornado with it. I, other developers, and possible future financial backers think that I have something new, and great in the making.

**Reply by stratika** [reply]
Nice, we are looking forward to see your progress. If you have any further questions or points to share feel free to append our discussion. 

Additionally, you can join the TornadoVM [Slack community](https://join.slack.com/t/tornadovmcommunity/shared_invite/zt-1sx0elmm8-hc~CUPO5xEW0vTaZKIXASg) if you wish to interact with more people and share your findings.
